File: IKSolver.pas
unit IKSolver;

{*******************************************************************************
  IKSolver - Inverse Kinematics Solution for Skeleton Joint Manipulation
  
  This unit provides IK algorithms for moving hands, arms, and fingers
  by calculating required joint rotations to reach target positions.
  
  Supports:
  - CCD (Cyclic Coordinate Descent) algorithm
  - FABRIK (Forward And Backward Reaching Inverse Kinematics) algorithm
  - Joint constraints and angle limits
*******************************************************************************}

interface

uses
  System.SysUtils, System.Math;

type
  TVector3D = record
    X, Y, Z: Double;
    // Vector operations
    class function Create(AX, AY, AZ: Double): TVector3D; static;
    function Add(const V: TVector3D): TVector3D;
    function Subtract(const V: TVector3D): TVector3D;
    function Multiply(Scalar: Double): TVector3D;
    function Divide(Scalar: Double): TVector3D;
    function Magnitude: Double;
    function Normalize: TVector3D;
    function Dot(const V: TVector3D): Double;
    function Cross(const V: TVector3D): TVector3D;
    function Distance(const V: TVector3D): Double;
  end;

  TJoint = record
    Position: TVector3D;
    MinAngle: Double;  // Constraint in degrees
    MaxAngle: Double;  // Constraint in degrees
    class function Create(const APosition: TVector3D; AMinAngle: Double = -180; AMaxAngle: Double = 180): TJoint; static;
  end;

  TJointArray = array of TJoint;

  TIKSolver = class
  private
    class function RotateAroundAxis(const Vector, Axis: TVector3D; Angle: Double): TVector3D;
  public
    { Solve IK using CCD (Cyclic Coordinate Descent) algorithm }
    class function SolveCCD(const Chain: TJointArray; const TargetPosition: TVector3D;
      Iterations: Integer = 10; Tolerance: Double = 0.01): TJointArray;
    
    { Solve IK using FABRIK (Forward And Backward Reaching Inverse Kinematics) algorithm }
    class function SolveFABRIK(const Chain: TJointArray; const TargetPosition: TVector3D;
      Iterations: Integer = 10; Tolerance: Double = 0.01): TJointArray;
    
    { Calculate angles between consecutive joints in a chain }
    class function CalculateJointAngles(const Chain: TJointArray): TArray<Double>;
    
    { Simple single joint IK - directly moves a joint to target position }
    class function SolveSingleJoint(const Joint: TJoint; const TargetPosition: TVector3D): TJoint;
    
    { Apply joint constraints to ensure angles are within limits }
    class function ApplyConstraints(const Chain: TJointArray): TJointArray;
    
    { Helper function to create a finger chain }
    class function CreateFingerChain(const BasePosition: TVector3D; 
      const BoneLengths: TArray<Double>; const Direction: TVector3D): TJointArray;
    
    { Helper function to create an arm chain }
    class function CreateArmChain(const ShoulderPosition: TVector3D;
      UpperArmLength: Double = 1.0; ForearmLength: Double = 1.0): TJointArray;
  end;

implementation

{ TVector3D }

class function TVector3D.Create(AX, AY, AZ: Double): TVector3D;
begin
  Result.X := AX;
  Result.Y := AY;
  Result.Z := AZ;
end;

function TVector3D.Add(const V: TVector3D): TVector3D;
begin
  Result.X := X + V.X;
  Result.Y := Y + V.Y;
  Result.Z := Z + V.Z;
end;

function TVector3D.Subtract(const V: TVector3D): TVector3D;
begin
  Result.X := X - V.X;
  Result.Y := Y - V.Y;
  Result.Z := Z - V.Z;
end;

function TVector3D.Multiply(Scalar: Double): TVector3D;
begin
  Result.X := X * Scalar;
  Result.Y := Y * Scalar;
  Result.Z := Z * Scalar;
end;

function TVector3D.Divide(Scalar: Double): TVector3D;
begin
  if Abs(Scalar) < 1e-10 then
  begin
    Result := TVector3D.Create(0, 0, 0);
    Exit;
  end;
  Result.X := X / Scalar;
  Result.Y := Y / Scalar;
  Result.Z := Z / Scalar;
end;

function TVector3D.Magnitude: Double;
begin
  Result := Sqrt(X * X + Y * Y + Z * Z);
end;

function TVector3D.Normalize: TVector3D;
var
  Mag: Double;
begin
  Mag := Magnitude;
  if Mag > 0 then
    Result := Divide(Mag)
  else
    Result := TVector3D.Create(0, 0, 0);
end;

function TVector3D.Dot(const V: TVector3D): Double;
begin
  Result := X * V.X + Y * V.Y + Z * V.Z;
end;

function TVector3D.Cross(const V: TVector3D): TVector3D;
begin
  Result.X := Y * V.Z - Z * V.Y;
  Result.Y := Z * V.X - X * V.Z;
  Result.Z := X * V.Y - Y * V.X;
end;

function TVector3D.Distance(const V: TVector3D): Double;
begin
  Result := Subtract(V).Magnitude;
end;

{ TJoint }

class function TJoint.Create(const APosition: TVector3D; AMinAngle, AMaxAngle: Double): TJoint;
begin
  Result.Position := APosition;
  Result.MinAngle := AMinAngle;
  Result.MaxAngle := AMaxAngle;
end;

{ TIKSolver }

class function TIKSolver.SolveCCD(const Chain: TJointArray; const TargetPosition: TVector3D;
  Iterations: Integer; Tolerance: Double): TJointArray;
var
  WorkChain: TJointArray;
  EndEffectorIndex: Integer;
  Iter, I, J: Integer;
  EndEffector: TVector3D;
  DistanceToTarget: Double;
  CurrentJoint, CurrentEnd: TVector3D;
  ToEnd, ToTarget, ToEndNorm, ToTargetNorm: TVector3D;
  DotProduct, Angle: Double;
  RotationAxis: TVector3D;
  RelPos, Rotated: TVector3D;
begin
  if Length(Chain) < 2 then
    raise Exception.Create('Chain must have at least 2 joints');

  // Clone the chain
  SetLength(WorkChain, Length(Chain));
  for I := 0 to High(Chain) do
    WorkChain[I] := Chain[I];

  EndEffectorIndex := High(WorkChain);

  for Iter := 0 to Iterations - 1 do
  begin
    // Check if we've reached the target
    EndEffector := WorkChain[EndEffectorIndex].Position;
    DistanceToTarget := EndEffector.Distance(TargetPosition);
    
    if DistanceToTarget < Tolerance then
      Break; // Solution found

    // Iterate backwards through the chain (excluding end effector)
    for I := EndEffectorIndex - 1 downto 0 do
    begin
      CurrentJoint := WorkChain[I].Position;
      CurrentEnd := WorkChain[EndEffectorIndex].Position;

      // Vector from current joint to end effector
      ToEnd := CurrentEnd.Subtract(CurrentJoint);
      // Vector from current joint to target
      ToTarget := TargetPosition.Subtract(CurrentJoint);

      // Calculate rotation angle
      ToEndNorm := ToEnd.Normalize;
      ToTargetNorm := ToTarget.Normalize;
      
      // Calculate angle between vectors
      DotProduct := Max(-1.0, Min(1.0, ToEndNorm.Dot(ToTargetNorm)));
      Angle := ArcCos(DotProduct);

      // Skip if angle is too small
      if Abs(Angle) < 0.001 then
        Continue;

      // Get rotation axis
      RotationAxis := ToEndNorm.Cross(ToTargetNorm).Normalize;

      // Apply rotation to all joints from current to end
      for J := I + 1 to EndEffectorIndex do
      begin
        RelPos := WorkChain[J].Position.Subtract(CurrentJoint);
        Rotated := RotateAroundAxis(RelPos, RotationAxis, Angle);
        WorkChain[J].Position := CurrentJoint.Add(Rotated);
      end;
    end;
  end;

  Result := WorkChain;
end;

class function TIKSolver.SolveFABRIK(const Chain: TJointArray; const TargetPosition: TVector3D;
  Iterations: Integer; Tolerance: Double): TJointArray;
var
  WorkChain: TJointArray;
  N, I, Iter: Integer;
  BoneLengths: TArray<Double>;
  RootPos, EndEffector, Direction: TVector3D;
  TotalLength, DistanceToTarget: Double;
  Offset: TVector3D;
begin
  if Length(Chain) < 2 then
    raise Exception.Create('Chain must have at least 2 joints');

  // Clone the chain
  SetLength(WorkChain, Length(Chain));
  for I := 0 to High(Chain) do
    WorkChain[I] := Chain[I];

  N := Length(WorkChain);
  
  // Store original bone lengths
  SetLength(BoneLengths, N - 1);
  for I := 0 to N - 2 do
    BoneLengths[I] := WorkChain[I].Position.Distance(WorkChain[I + 1].Position);

  // Store the root position
  RootPos := WorkChain[0].Position;

  // Calculate total chain length
  TotalLength := 0;
  for I := 0 to High(BoneLengths) do
    TotalLength := TotalLength + BoneLengths[I];

  DistanceToTarget := RootPos.Distance(TargetPosition);

  // If target is unreachable, stretch chain towards target
  if DistanceToTarget > TotalLength then
  begin
    Direction := TargetPosition.Subtract(RootPos).Normalize;
    for I := 1 to N - 1 do
    begin
      Offset := Direction.Multiply(BoneLengths[I - 1]);
      WorkChain[I].Position := WorkChain[I - 1].Position.Add(Offset);
    end;
    Result := WorkChain;
    Exit;
  end;

  // FABRIK iterations
  for Iter := 0 to Iterations - 1 do
  begin
    // Check convergence
    EndEffector := WorkChain[N - 1].Position;
    if EndEffector.Distance(TargetPosition) < Tolerance then
      Break;

    // Forward reaching phase
    WorkChain[N - 1].Position := TargetPosition;
    for I := N - 2 downto 0 do
    begin
      Direction := WorkChain[I].Position.Subtract(WorkChain[I + 1].Position).Normalize;
      WorkChain[I].Position := WorkChain[I + 1].Position.Add(Direction.Multiply(BoneLengths[I]));
    end;

    // Backward reaching phase
    WorkChain[0].Position := RootPos;
    for I := 0 to N - 2 do
    begin
      Direction := WorkChain[I + 1].Position.Subtract(WorkChain[I].Position).Normalize;
      WorkChain[I + 1].Position := WorkChain[I].Position.Add(Direction.Multiply(BoneLengths[I]));
    end;
  end;

  Result := WorkChain;
end;

class function TIKSolver.RotateAroundAxis(const Vector, Axis: TVector3D; Angle: Double): TVector3D;
var
  CosAngle, SinAngle: Double;
  Term1, Term2, Term3: TVector3D;
begin
  CosAngle := Cos(Angle);
  SinAngle := Sin(Angle);

  // Rodrigues' rotation formula
  Term1 := Vector.Multiply(CosAngle);
  Term2 := Axis.Cross(Vector).Multiply(SinAngle);
  Term3 := Axis.Multiply(Axis.Dot(Vector) * (1 - CosAngle));

  Result := Term1.Add(Term2).Add(Term3);
end;

class function TIKSolver.CalculateJointAngles(const Chain: TJointArray): TArray<Double>;
var
  I: Integer;
  V1, V2: TVector3D;
  DotProduct, Angle: Double;
begin
  SetLength(Result, Length(Chain) - 2);
  for I := 1 to Length(Chain) - 2 do
  begin
    V1 := Chain[I].Position.Subtract(Chain[I - 1].Position).Normalize;
    V2 := Chain[I + 1].Position.Subtract(Chain[I].Position).Normalize;
    DotProduct := Max(-1.0, Min(1.0, V1.Dot(V2)));
    Angle := ArcCos(DotProduct) * (180 / Pi);
    Result[I - 1] := Angle;
  end;
end;

class function TIKSolver.SolveSingleJoint(const Joint: TJoint; const TargetPosition: TVector3D): TJoint;
begin
  Result := Joint;
  Result.Position := TargetPosition;
end;

class function TIKSolver.ApplyConstraints(const Chain: TJointArray): TJointArray;
var
  I: Integer;
  V1, V2: TVector3D;
  CurrentAngle, TargetAngle: Double;
begin
  SetLength(Result, Length(Chain));
  for I := 0 to High(Chain) do
    Result[I] := Chain[I];

  for I := 1 to Length(Chain) - 2 do
  begin
    V1 := Result[I].Position.Subtract(Result[I - 1].Position);
    V2 := Result[I + 1].Position.Subtract(Result[I].Position);
    
    CurrentAngle := ArcCos(Max(-1.0, Min(1.0, V1.Normalize.Dot(V2.Normalize)))) * (180 / Pi);

    // Clamp angle within constraints
    if (CurrentAngle < Result[I].MinAngle) or (CurrentAngle > Result[I].MaxAngle) then
    begin
      TargetAngle := Max(Result[I].MinAngle, Min(Result[I].MaxAngle, CurrentAngle));
      // Note: Actual constraint adjustment would require more complex calculations
      // This is a simplified version for demonstration
    end;
  end;
end;

class function TIKSolver.CreateFingerChain(const BasePosition: TVector3D; 
  const BoneLengths: TArray<Double>; const Direction: TVector3D): TJointArray;
var
  I: Integer;
  Dir: TVector3D;
  CurrentPos: TVector3D;
begin
  SetLength(Result, Length(BoneLengths) + 1);
  Result[0] := TJoint.Create(BasePosition);
  
  Dir := Direction.Normalize;
  CurrentPos := BasePosition;
  
  for I := 0 to High(BoneLengths) do
  begin
    CurrentPos := CurrentPos.Add(Dir.Multiply(BoneLengths[I]));
    Result[I + 1] := TJoint.Create(CurrentPos, -20, 110); // Typical finger constraints
  end;
end;

class function TIKSolver.CreateArmChain(const ShoulderPosition: TVector3D;
  UpperArmLength, ForearmLength: Double): TJointArray;
begin
  SetLength(Result, 3);
  Result[0] := TJoint.Create(ShoulderPosition, -180, 180);
  Result[1] := TJoint.Create(ShoulderPosition.Add(TVector3D.Create(0, -UpperArmLength, 0)), 0, 160);
  Result[2] := TJoint.Create(ShoulderPosition.Add(TVector3D.Create(0, -(UpperArmLength + ForearmLength), 0)), -90, 90);
end;

end.

File: IKSolver.js
/**
 * IKSolver.js - Inverse Kinematics Solution for Skeleton Joint Manipulation
 * 
 * This module provides IK algorithms for moving hands, arms, and fingers
 * by calculating required joint rotations to reach target positions.
 * 
 * Supports:
 * - CCD (Cyclic Coordinate Descent) algorithm
 * - FABRIK (Forward And Backward Reaching Inverse Kinematics) algorithm
 * - Joint constraints and angle limits
 */

class Vector3D {
    constructor(x = 0, y = 0, z = 0) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    // Vector operations
    add(v) {
        return new Vector3D(this.x + v.x, this.y + v.y, this.z + v.z);
    }

    subtract(v) {
        return new Vector3D(this.x - v.x, this.y - v.y, this.z - v.z);
    }

    multiply(scalar) {
        return new Vector3D(this.x * scalar, this.y * scalar, this.z * scalar);
    }

    divide(scalar) {
        if (scalar === 0) return new Vector3D(0, 0, 0);
        return new Vector3D(this.x / scalar, this.y / scalar, this.z / scalar);
    }

    magnitude() {
        return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
    }

    normalize() {
        const mag = this.magnitude();
        return mag > 0 ? this.divide(mag) : new Vector3D(0, 0, 0);
    }

    dot(v) {
        return this.x * v.x + this.y * v.y + this.z * v.z;
    }

    cross(v) {
        return new Vector3D(
            this.y * v.z - this.z * v.y,
            this.z * v.x - this.x * v.z,
            this.x * v.y - this.y * v.x
        );
    }

    distance(v) {
        return this.subtract(v).magnitude();
    }

    clone() {
        return new Vector3D(this.x, this.y, this.z);
    }
}

class Joint {
    constructor(position, minAngle = -180, maxAngle = 180) {
        this.position = position instanceof Vector3D ? position : new Vector3D(position.x, position.y, position.z);
        this.minAngle = minAngle; // Constraint in degrees
        this.maxAngle = maxAngle; // Constraint in degrees
        this.rotation = new Vector3D(0, 0, 0); // Euler angles in degrees (x, y, z)
    }

    clone() {
        const joint = new Joint(this.position.clone(), this.minAngle, this.maxAngle);
        joint.rotation = this.rotation.clone();
        return joint;
    }
}

class IKSolver {
    /**
     * Solve IK using CCD (Cyclic Coordinate Descent) algorithm
     * @param {Array<Joint>} chain - Array of joints from root to end effector
     * @param {Vector3D} targetPosition - Desired position for end effector
     * @param {Number} iterations - Maximum iterations (default: 10)
     * @param {Number} tolerance - Distance tolerance for convergence (default: 0.01)
     * @returns {Array<Joint>} - Updated joint chain
     */
    static solveCCD(chain, targetPosition, iterations = 10, tolerance = 0.01) {
        if (!chain || chain.length < 2) {
            throw new Error("Chain must have at least 2 joints");
        }

        // Clone the chain to avoid modifying the original
        const workChain = chain.map(joint => joint.clone());
        const endEffectorIndex = workChain.length - 1;

        for (let iter = 0; iter < iterations; iter++) {
            // Check if we've reached the target
            const endEffector = workChain[endEffectorIndex].position;
            const distanceToTarget = endEffector.distance(targetPosition);
            
            if (distanceToTarget < tolerance) {
                break; // Solution found
            }

            // Iterate backwards through the chain (excluding end effector)
            for (let i = endEffectorIndex - 1; i >= 0; i--) {
                const currentJoint = workChain[i].position;
                const currentEnd = workChain[endEffectorIndex].position;

                // Vector from current joint to end effector
                const toEnd = currentEnd.subtract(currentJoint);
                // Vector from current joint to target
                const toTarget = targetPosition.subtract(currentJoint);

                // Calculate rotation angle
                const toEndNorm = toEnd.normalize();
                const toTargetNorm = toTarget.normalize();
                
                // Calculate angle between vectors
                const dotProduct = Math.max(-1, Math.min(1, toEndNorm.dot(toTargetNorm)));
                const angle = Math.acos(dotProduct);

                // Skip if angle is too small
                if (Math.abs(angle) < 0.001) continue;

                // Get rotation axis
                const rotationAxis = toEndNorm.cross(toTargetNorm).normalize();

                // Apply rotation to all joints from current to end
                for (let j = i + 1; j <= endEffectorIndex; j++) {
                    const relPos = workChain[j].position.subtract(currentJoint);
                    const rotated = this.rotateAroundAxis(relPos, rotationAxis, angle);
                    workChain[j].position = currentJoint.add(rotated);
                }
            }
        }

        // Calculate and store bone rotations
        return this.calculateBoneRotations(workChain);
    }

    /**
     * Solve IK using FABRIK (Forward And Backward Reaching Inverse Kinematics) algorithm
     * @param {Array<Joint>} chain - Array of joints from root to end effector
     * @param {Vector3D} targetPosition - Desired position for end effector
     * @param {Number} iterations - Maximum iterations (default: 10)
     * @param {Number} tolerance - Distance tolerance for convergence (default: 0.01)
     * @returns {Array<Joint>} - Updated joint chain
     */
    static solveFABRIK(chain, targetPosition, iterations = 10, tolerance = 0.01) {
        if (!chain || chain.length < 2) {
            throw new Error("Chain must have at least 2 joints");
        }

        // Clone the chain
        const workChain = chain.map(joint => joint.clone());
        const n = workChain.length;
        
        // Store original bone lengths
        const boneLengths = [];
        for (let i = 0; i < n - 1; i++) {
            boneLengths[i] = workChain[i].position.distance(workChain[i + 1].position);
        }

        // Store the root position
        const rootPos = workChain[0].position.clone();

        // Calculate total chain length
        const totalLength = boneLengths.reduce((sum, length) => sum + length, 0);
        const distanceToTarget = rootPos.distance(targetPosition);

        // If target is unreachable, stretch chain towards target
        if (distanceToTarget > totalLength) {
            const direction = targetPosition.subtract(rootPos).normalize();
            for (let i = 1; i < n; i++) {
                const offset = direction.multiply(boneLengths[i - 1]);
                workChain[i].position = workChain[i - 1].position.add(offset);
            }
            return workChain;
        }

        // FABRIK iterations
        for (let iter = 0; iter < iterations; iter++) {
            // Check convergence
            const endEffector = workChain[n - 1].position;
            if (endEffector.distance(targetPosition) < tolerance) {
                break;
            }

            // Forward reaching phase
            workChain[n - 1].position = targetPosition.clone();
            for (let i = n - 2; i >= 0; i--) {
                const direction = workChain[i].position.subtract(workChain[i + 1].position).normalize();
                workChain[i].position = workChain[i + 1].position.add(direction.multiply(boneLengths[i]));
            }

            // Backward reaching phase
            workChain[0].position = rootPos.clone();
            for (let i = 0; i < n - 1; i++) {
                const direction = workChain[i + 1].position.subtract(workChain[i].position).normalize();
                workChain[i + 1].position = workChain[i].position.add(direction.multiply(boneLengths[i]));
            }
        }

        // Calculate and store bone rotations
        return this.calculateBoneRotations(workChain);
    }

    /**
     * Rotate a vector around an axis by an angle
     * Uses Rodrigues' rotation formula
     * @param {Vector3D} vector - Vector to rotate
     * @param {Vector3D} axis - Rotation axis (should be normalized)
     * @param {Number} angle - Rotation angle in radians
     * @returns {Vector3D} - Rotated vector
     */
    static rotateAroundAxis(vector, axis, angle) {
        const cosAngle = Math.cos(angle);
        const sinAngle = Math.sin(angle);

        // Rodrigues' rotation formula
        const term1 = vector.multiply(cosAngle);
        const term2 = axis.cross(vector).multiply(sinAngle);
        const term3 = axis.multiply(axis.dot(vector) * (1 - cosAngle));

        return term1.add(term2).add(term3);
    }

    /**
     * Calculate angles between consecutive joints in a chain
     * @param {Array<Joint>} chain - Joint chain
     * @returns {Array<Number>} - Array of angles in degrees
     */
    static calculateJointAngles(chain) {
        const angles = [];
        for (let i = 1; i < chain.length - 1; i++) {
            const v1 = chain[i].position.subtract(chain[i - 1].position).normalize();
            const v2 = chain[i + 1].position.subtract(chain[i].position).normalize();
            const dotProduct = Math.max(-1, Math.min(1, v1.dot(v2)));
            const angle = Math.acos(dotProduct) * (180 / Math.PI);
            angles.push(angle);
        }
        return angles;
    }

    /**
     * Simple single joint IK - directly moves a joint to target position
     * @param {Joint} joint - The joint to move
     * @param {Vector3D} targetPosition - Desired position
     * @returns {Joint} - Updated joint
     */
    static solveSingleJoint(joint, targetPosition) {
        const updatedJoint = joint.clone();
        updatedJoint.position = targetPosition.clone();
        return updatedJoint;
    }

    /**
     * Apply joint constraints to ensure angles are within limits
     * @param {Array<Joint>} chain - Joint chain
     * @returns {Array<Joint>} - Constrained chain
     */
    static applyConstraints(chain) {
        const constrainedChain = chain.map(joint => joint.clone());
        
        for (let i = 1; i < constrainedChain.length - 1; i++) {
            const v1 = constrainedChain[i].position.subtract(constrainedChain[i - 1].position);
            const v2 = constrainedChain[i + 1].position.subtract(constrainedChain[i].position);
            
            const currentAngle = Math.acos(
                Math.max(-1, Math.min(1, v1.normalize().dot(v2.normalize())))
            ) * (180 / Math.PI);

            const joint = constrainedChain[i];
            
            // Clamp angle within constraints
            if (currentAngle < joint.minAngle || currentAngle > joint.maxAngle) {
                const targetAngle = Math.max(joint.minAngle, Math.min(joint.maxAngle, currentAngle));
                // Adjust position to respect constraint
                // This is a simplified constraint - real implementation would be more complex
                console.warn(`Joint ${i} angle ${currentAngle}° outside limits [${joint.minAngle}°, ${joint.maxAngle}°]`);
            }
        }
        
        return constrainedChain;
    }

    /**
     * Calculate bone rotations (Euler angles) from joint positions
     * @param {Array<Joint>} chain - Joint chain with positions
     * @returns {Array<Joint>} - Chain with updated rotation values (in degrees)
     */
    static calculateBoneRotations(chain) {
        const rotatedChain = chain.map(joint => joint.clone());
        
        // For each bone (segment between two joints)
        for (let i = 0; i < rotatedChain.length - 1; i++) {
            const boneVector = rotatedChain[i + 1].position.subtract(rotatedChain[i].position).normalize();
            
            // Calculate Euler angles (ZYX convention)
            // Rotation around Y-axis (yaw)
            const yaw = Math.atan2(boneVector.x, boneVector.z) * (180 / Math.PI);
            
            // Rotation around X-axis (pitch)
            const pitch = Math.asin(-boneVector.y) * (180 / Math.PI);
            
            // For simplicity, roll (rotation around Z-axis) is set to 0
            // A more complex implementation would track twist along the bone
            const roll = 0;
            
            rotatedChain[i].rotation = new Vector3D(pitch, yaw, roll);
        }
        
        // Last joint's rotation is same as the previous bone's direction
        if (rotatedChain.length > 1) {
            rotatedChain[rotatedChain.length - 1].rotation = 
                rotatedChain[rotatedChain.length - 2].rotation.clone();
        }
        
        return rotatedChain;
    }

    /**
     * Helper function to create a finger chain (5 joints typical for fingers)
     * @param {Vector3D} basePosition - Starting position of finger
     * @param {Array<Number>} boneLengths - Array of bone lengths [3-4 elements]
     * @param {Vector3D} direction - Initial direction of finger
     * @returns {Array<Joint>} - Finger joint chain
     */
    static createFingerChain(basePosition, boneLengths = [1, 0.8, 0.6], direction = new Vector3D(1, 0, 0)) {
        const chain = [new Joint(basePosition.clone())];
        const dir = direction.normalize();
        
        let currentPos = basePosition.clone();
        for (const length of boneLengths) {
            currentPos = currentPos.add(dir.multiply(length));
            chain.push(new Joint(currentPos.clone(), -20, 110)); // Typical finger constraints
        }
        
        return chain;
    }

    /**
     * Helper function to create an arm chain (shoulder, elbow, wrist)
     * @param {Vector3D} shoulderPosition - Shoulder position
     * @param {Number} upperArmLength - Length of upper arm
     * @param {Number} forearmLength - Length of forearm
     * @returns {Array<Joint>} - Arm joint chain
     */
    static createArmChain(shoulderPosition, upperArmLength = 1.0, forearmLength = 1.0) {
        return [
            new Joint(shoulderPosition.clone(), -180, 180),
            new Joint(shoulderPosition.add(new Vector3D(0, -upperArmLength, 0)), 0, 160),
            new Joint(shoulderPosition.add(new Vector3D(0, -(upperArmLength + forearmLength), 0)), -90, 90)
        ];
    }
}

// Export for use in Node.js or browser
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { IKSolver, Vector3D, Joint };
}
